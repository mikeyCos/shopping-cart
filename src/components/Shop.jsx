import { Outlet, useRouteLoaderData } from "react-router-dom";
import Categories from "./Categories";
import styles from "../styles/Shop.module.css";

const Shop = () => {
  const { categories, products } = useRouteLoaderData("root");
  return (
    <section id={styles["shop"]}>
      <Categories categories={categories} />
      <Outlet context={[products]} />
    </section>
  );
};

export default Shop;

/*
  - Shop
    - Make use of https://fakestoreapi.com/
    - Individual card elements for each product
      - Is it possible to have clickable images that render a modal?
        - Each images is wrapped around a Link component
        - This will need a route
        - The modal's content is an outlet that renders the information of the product
      - Input quantity (type number)
        - (OPTIONAL), create custom increment and decrement buttons
        - Need to hide the inner-spin-button
        - https://stackoverflow.com/questions/75879418/how-to-remove-arrows-in-input-type-number-inside-the-input
      - Title
      - "Add To Cart" button

  - Going to the shop page
    - Creates a list of categories
      - Need to fetch categories from API
        - Should only fetch when shop page is loaded(?)
        - https://fakestoreapi.com/products/categories
      - These are links
      - This should always be visible on the shop page
      - "electronics", "jewelery", "men's clothing", "women's clothing"
      - Clicking "jewelery" sends path from /shop/ to /shop/jewelery
        - Fetch https://fakestoreapi.com/products/category/${category}
          - In this example, category = 'jewelery'
        - The path should now be /shop/category/${category}
    - Index should be the all products

  - When does it make sense to fetch data from an API?
    - If I fetch from an API on load, when the application/website first loads
      - The application will need to use more resources on load
      - All data is fetched one time
    - If I fetch from an API only when it is needed
      - The application will sparingly use resources
      - Data is fetched more than one time
*/
