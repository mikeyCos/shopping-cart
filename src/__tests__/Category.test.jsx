import { describe, expect, it, vi } from "vitest";
import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { MemoryRouter, Route, Routes, useParams } from "react-router-dom";
import { categories, parsedProducts, products } from "./data.mocks";
import Shop from "../components/Shop";

const fetchMock = vi.fn(() => {
  return Promise.resolve({
    json: () =>
      Promise.resolve(
        vi.fn().mockReturnValueOnce(categories).mockReturnValueOnce(products)
      ),
  });
});

vi.stubGlobal("fetch", fetchMock);
// Why and how does mocking the module fix testing issues?
vi.mock("../utilities/parseProducts");

const Category = () => {
  const { category = "all" } = useParams();
  return (
    <section>
      <h2>Category</h2>
      {category && <h3>{category}</h3>}
      {products && (
        <section role="region">
          {parsedProducts[category].map((product) => (
            <article key={product.id}>
              <picture>
                <img src={product.image} />
              </picture>
              <div className="info">
                <h4>{product.title}</h4>
              </div>
            </article>
          ))}
        </section>
      )}
    </section>
  );
};

describe("Category component (Outlet)", () => {
  // How to test Outlet component?
  // https://stackoverflow.com/questions/70654872/how-to-test-react-router-v6-outlet-using-testing-library-react
  it("Category component is rendered", async () => {
    window.history.pushState({}, "Shop", "/shop/");
    const { container } = render(
      // <MemoryRouter initialEntries={["/shop"]}>
      //   <Routes>
      //     <Route path="/shop" element={<Shop />}>
      //       <Route path="/shop" element={<Category />} />
      //       <Route path="/shop/:category" element={<Category />} />
      //     </Route>
      //   </Routes>
      // </MemoryRouter>
      <MemoryRouter initialEntries={[`/shop`]}>
        <Routes>
          <Route path="/shop" element={<Shop />}>
            <Route path="/shop/" element={<Category />} />
          </Route>
        </Routes>
      </MemoryRouter>
    );

    await screen.findByRole("heading", { level: 3 });
    expect(container).matchSnapshot();
  });

  it("Category's heading text should be 'electronics'", async () => {
    const category = "electronics";
    const regExp = new RegExp(category); // regular expression: /electronics/
    render(
      <MemoryRouter initialEntries={[`/shop/category/${category}`]}>
        <Routes>
          <Route path="/shop" element={<Shop />}>
            <Route path="/shop" element={<Category />} />
            <Route path="/shop/category/:category" element={<Category />} />
          </Route>
        </Routes>
      </MemoryRouter>
    );

    const categoryElectronics = await screen.findByRole("heading", {
      level: 3,
    });
    expect(categoryElectronics.textContent).toMatch(regExp);
  });

  it("20 products are listed for the 'all' category", async () => {
    const allProducts = parsedProducts["all"]; // length of 20
    render(
      <MemoryRouter initialEntries={[`/shop`]}>
        <Routes>
          <Route path="/shop" element={<Shop />}>
            <Route path="/shop" element={<Category />} />
            <Route path="/shop/category/:category" element={<Category />} />
          </Route>
        </Routes>
      </MemoryRouter>
    );

    const productsList = await screen.findByRole("region");
    expect(productsList.children.length).toBe(allProducts.length);
  });

  it("20 products are listed for the 'electronics' category", async () => {
    const category = "electronics";
    const electronics = parsedProducts[category]; // length of 6
    render(
      <MemoryRouter initialEntries={[`/shop/category/${category}`]}>
        <Routes>
          <Route path="/shop" element={<Shop />}>
            <Route path="/shop" element={<Category />} />
            <Route path="/shop/category/:category" element={<Category />} />
          </Route>
        </Routes>
      </MemoryRouter>
    );

    const electronicsList = await screen.findByRole("region");
    expect(electronicsList.children.length).toBe(electronics.length);
  });
});
