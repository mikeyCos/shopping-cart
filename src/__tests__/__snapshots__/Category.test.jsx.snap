// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Category component (Outlet) > Category component is rendered 1`] = `
<div>
  <section>
    <h2>
      Category
    </h2>
    <h3>
      all
    </h3>
    <section
      role="region"
    >
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            Mens Casual Premium Slim Fit T-Shirts 
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            Mens Cotton Jacket
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/71YXzeOuslL._AC_UY879_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            Mens Casual Slim Fit
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/71pWzhdJNwL._AC_UL640_QL65_ML3_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            John Hardy Women's Legends Naga Gold & Silver Dragon Station Chain Bracelet
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/61sbMiUnoGL._AC_UL640_QL65_ML3_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            Solid Gold Petite Micropave 
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/71YAIFU48IL._AC_UL640_QL65_ML3_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            White Gold Plated Princess
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/51UDEzMJVpL._AC_UL640_QL65_ML3_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            Pierced Owl Rose Gold Plated Stainless Steel Double
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/61IBBVJvSDL._AC_SY879_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            WD 2TB Elements Portable External Hard Drive - USB 3.0 
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/61U7T1koQqL._AC_SX679_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            SanDisk SSD PLUS 1TB Internal SSD - SATA III 6 Gb/s
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/71kWymZ+c+L._AC_SX679_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            Silicon Power 256GB SSD 3D NAND A55 SLC Cache Performance Boost SATA III 2.5
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/61mtL65D4cL._AC_SX679_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            WD 4TB Gaming Drive Works with Playstation 4 Portable External Hard Drive
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/81QpkIctqPL._AC_SX679_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            Acer SB220Q bi 21.5 inches Full HD (1920 x 1080) IPS Ultra-Thin
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/81Zt42ioCgL._AC_SX679_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            Samsung 49-Inch CHG90 144Hz Curved Gaming Monitor (LC49HG90DMNXZA) â€“ Super Ultrawide Screen QLED 
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/51Y5NI-I5jL._AC_UX679_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            BIYLACLESEN Women's 3-in-1 Snowboard Jacket Winter Coats
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/81XH0e8fefL._AC_UY879_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            Lock and Love Women's Removable Hooded Faux Leather Moto Biker Jacket
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/71HblAHs5xL._AC_UY879_-2.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            Rain Jacket Women Windbreaker Striped Climbing Raincoats
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/71z3kpMAYsL._AC_UY879_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            MBJ Women's Solid Short Sleeve Boat Neck V 
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/51eg55uWmdL._AC_UX679_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            Opna Women's Short Sleeve Moisture
          </h4>
        </div>
      </article>
      <article>
        <picture>
          <img
            src="https://fakestoreapi.com/img/61pHAEJ4NML._AC_UX679_.jpg"
          />
        </picture>
        <div
          class="info"
        >
          <h4>
            DANVOUY Womens T Shirt Casual Cotton Short
          </h4>
        </div>
      </article>
    </section>
  </section>
</div>
`;
